src/main/main.c

```diff
@@ -804,47 +878,52 @@ static int fd_poll(struct re *re)
 	if (n < 0)
 		return ERRNO_SOCK;
 
-	/* Check for events */
-	for (i=0; (n > 0) && (i < re->nfds); i++) {
+	for (int i = 0; (n > 0) && (i < nfds); i++) {
 		re_sock_t fd;
+		struct fhs *fhs = NULL;
 		int flags = 0;
 
 		switch (re->method) {
-
+#ifdef HAVE_SELECT
+		case METHOD_SELECT:
+#ifdef WIN32
+			fd = sfds[i];
+#else
+			fd = i;
+#endif
+			if (FD_ISSET(fd, &rfds))
+				flags |= FD_READ;
+			if (FD_ISSET(fd, &wfds))
+				flags |= FD_WRITE;
+			if (FD_ISSET(fd, &efds))
+				flags |= FD_EXCEPT;
+			break;
+#endif
 #ifdef HAVE_POLL
 		case METHOD_POLL:
-			fd = i;
-			if (re->fds[fd].revents & POLLIN)
+			fd = re->fds[i].fd;
+
+			if (re->fds[i].revents & POLLIN)
 				flags |= FD_READ;
-			if (re->fds[fd].revents & POLLOUT)
+			if (re->fds[i].revents & POLLOUT)
 				flags |= FD_WRITE;
-			if (re->fds[fd].revents & (POLLERR|POLLHUP|POLLNVAL))
+			if (re->fds[i].revents & (POLLERR|POLLHUP|POLLNVAL))
 				flags |= FD_EXCEPT;
-			if (re->fds[fd].revents & POLLNVAL) {
+			if (re->fds[i].revents & POLLNVAL) {
 				DEBUG_WARNING("event: fd=%d POLLNVAL"
 					      " (fds.fd=%d,"
 					      " fds.events=0x%02x)\n",
-					      fd, re->fds[fd].fd,
-					      re->fds[fd].events);
+					      fd, re->fds[i].fd,
+					      re->fds[i].events);
 			}
 			/* Clear events */
-			re->fds[fd].revents = 0;
-			break;
-#endif
-#ifdef HAVE_SELECT
-		case METHOD_SELECT:
-			fd = re->fhs[i].fd;
-			if (FD_ISSET(fd, &rfds))
-				flags |= FD_READ;
-			if (FD_ISSET(fd, &wfds))
-				flags |= FD_WRITE;
-			if (FD_ISSET(fd, &efds))
-				flags |= FD_EXCEPT;
+			re->fds[i].revents = 0;
 			break;
 #endif
 #ifdef HAVE_EPOLL
 		case METHOD_EPOLL:
-			fd = re->events[i].data.fd;
+			fhs = re->events[i].data.ptr;
```

Comment:
This is vulnerable to a race with another thread calling `fd_close`:

1. Some thread calls `fd_listen(fd)`.
2. Event loop goes into `epoll_wait`, releasing the `re` lock.
3. An event is triggered on `fd`.
4. Another thread calls `fd_close(fd)` and removes `fhs` from the hash table and deletes it.
5. Event loop wakes up and receives the event on the `fd`. (Alternatively, it wakes up during step 4 and blocks on `re` lock.)
6. Here, `data.ptr` is invalid.

So, `epoll` backend needs to perform a lookup in the hash table for `fhs`.

But that's not all. The `epoll` implementation needs to account not only for such races but also for the ABA problem: one thread may remove and close fd from event loop and then add a *different* fd *with the same value* before the event loop thread manages to lock `re` and get to processing the event on the *old* fd. The implementation has to use fd and an ABA counter in the `epoll_event::data.u64` member to mitigate this. The ABA counter needs to be maintained in `re` context.

Reply:
Would marking `fhs` hash entry only for delete (put on a separate list) and delete after processing all events prevent both issues?

Reply:
It wouldn't help against the ABA problem. Since after `fd_close` the fd is reused, you would have the same fd value both in the delete list and `fhs`. And that replacement could potentially happen multiple times, so there may be multiple entries for the same fd in the delete list.

If you think this problem is unlikely to happen, I can assure you it can, and of course at the most unfortunate moment too. Ask me how I know. :)

Reply:
Though, on the second thought, if by "marking" you mean changing its contents so that the event loop can tell stale `fhs` entries from the actual ones then yes, it could work.

Reply:
I meant something like this in fd_listen:
```c
	if (!flags) {
		list_append(re->fhs_delete, &fhs->le, fhs);
		fhs->index = -1;
		--re->nfds;
	}
```

And this check?
```c
		if (fhs->fh && fhs->index >= 0) {
```

Reply:
Yes, something like this could work.

Reply:
Ok I refactored this